src/AddAtomicMutex.cpp

```diff
@@ -0,0 +1,266 @@
+public:
+    using IRMutator::visit;
+
+    const Target &target;
+
+    AddAtomicMutex(const Target &target) : target(target) {}
+
+    Stmt visit(const ProducerConsumer *op) override {
+        FindAtomicMutexUsage finder(op->name);
+        if (op->is_producer) {
+            finder.visit(op);
+            if (!finder.found) {
+                return IRMutator::visit(op);
+            }
+        } else {
+            return IRMutator::visit(op);
+        }
+
+        const std::string &producer_name = finder.producer_name;
+        const std::string &mutex_name = finder.mutex_name;
+        int tuple_size = finder.tuple_size;
+        int dimensions = finder.dimensions;
+        const char *extent_field = tuple_size == 1 ? ".extent." : ".0.extent.";
+        const char *min_field = tuple_size == 1 ? ".min." : ".0.min.";
+        const char *stride_field = tuple_size == 1 ? ".stride." : ".0.stride.";
+        Expr buffer_size = Expr(1);
+        for (int i = 0; i < dimensions; i++) {
+            Expr extent = Variable::make(Int(32), producer_name + extent_field + std::to_string(i));
+            buffer_size *= extent;
+        }
+        Expr mutex_array = Call::make(type_of<halide_mutex_array **>(),
+                                      "halide_mutex_array_create",
+                                      {buffer_size},
+                                      Call::Extern);
+        Stmt body = mutate(op->body);
+        // Allocate a scalar of halide_mutex_array.
+        // This generate halide_mutex_array *mutex[1];
+        body = Allocate::make(mutex_name,
+                              type_of<halide_mutex_array *>(),
+                              MemoryType::Stack,
+                              {},
+                              const_true(),
+                              body,
+                              mutex_array,
+                              "halide_mutex_array_destroy");
+        // Insert Let statements for the buffer's min and stride for storage flattening.
+        for (int i = 0; i < dimensions; i++) {
+            body = LetStmt::make(mutex_name + ".min." + std::to_string(i),
+                                 Variable::make(Int(32), producer_name + min_field + std::to_string(i)),
+                                 body);
+            body = LetStmt::make(mutex_name + ".stride." + std::to_string(i),
+                                 Variable::make(Int(32), producer_name + stride_field + std::to_string(i)),
+                                 body);
+        }
+
+        return ProducerConsumer::make(op->name, op->is_producer, std::move(body));
+    }
+
+    Stmt visit(const Atomic *op) override {
+        if (op->mutex_name == "") {
+            return IRMutator::visit(op);
+        }
+
+        // op->mutex_indices.size() could be 0.
+        Expr index = op->mutex_indices.size() == 1 ? op->mutex_indices[0] : Expr(0);
+        Stmt body = op->body;
+        Expr mutex = Load::make(type_of<halide_mutex_array *>(),
```

Comment:
I think you can avoid the double-pointer. An Allocate node of type T with name "foo" lets you make load nodes from foo of type T, but it also places a Variable "foo" of type T* in scope. So you don't need to have a load node here - you can just make a Variable and pass it to the lock/unlock calls.

Reply:
Ah great idea let me do that

Reply:
Done.